File: src/main/java/hu/frontrider/arcana/capabilities/CreatureEnchantCapability.java
package hu.frontrider.arcana.capabilities;

import hu.frontrider.arcana.creatureenchant.backend.CreatureEnchant;
import hu.frontrider.arcana.creatureenchant.backend.EnchantingBaseCircle;
import net.minecraftforge.fml.common.registry.GameRegistry;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

import static hu.frontrider.arcana.ThaumicArcana.MODID;

public class CreatureEnchantCapability implements ICreatureEnchant {

    @GameRegistry.ObjectHolder(MODID + ":normal")
    private static EnchantingBaseCircle baseCircle = null;

    private Map<CreatureEnchant, CreatureEnchantContainer> enchants;
    private EnchantingBaseCircle enchantingBaseCircle;

    public CreatureEnchantCapability() {
        enchants = new HashMap<>();
        enchantingBaseCircle = baseCircle;
    }

    @Override
    public boolean hasEnchant() {
        return !enchants.isEmpty();
    }


    @Override
    public boolean hasEnchant(CreatureEnchant enchantment) {
        return enchants.containsKey(enchantment);
    }

    @Override
    public int getLevel(CreatureEnchant enchantment) {
        return enchants.get(enchantment) != null ? enchants.get(enchantment).level : 0;
    }

    @Override
    public Map<CreatureEnchant, CreatureEnchantContainer> getStore() {
        return enchants;
    }

    @Override
    public void setStore(Map<CreatureEnchant, CreatureEnchantContainer> store) {
        enchants = store;
    }

    @Override
    public void putEnchant(CreatureEnchantContainer enchantment) {
        enchants.put(enchantment.creatureEnchant,enchantment);
    }

    @Override
    public EnchantingBaseCircle getCircle() {
        return enchantingBaseCircle;
    }

    @Override
    public void setCircle(EnchantingBaseCircle enchantingBaseCircle) {
        this.enchantingBaseCircle = enchantingBaseCircle;
    }

    public static class CreatureEnchantContainer {
        private final CreatureEnchant creatureEnchant;
        private final int level;
        private final int usedTo;

        public CreatureEnchantContainer(CreatureEnchant creatureEnchant, int level, int usedTo){
            this.creatureEnchant = creatureEnchant;
            this.level = level;
            this.usedTo = usedTo;
        }

        public CreatureEnchant getCreatureEnchant() {
            return creatureEnchant;
        }

        public int getLevel() {
            return level;
        }

        public int getUsedTo() {
            return usedTo;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            CreatureEnchantContainer that = (CreatureEnchantContainer) o;
            return level == that.level &&
                    usedTo == that.usedTo &&
                    Objects.equals(creatureEnchant, that.creatureEnchant);
        }

        @Override
        public int hashCode() {

            return Objects.hash(creatureEnchant, level, usedTo);
        }

        @Override
        public String toString() {
            return "CreatureEnchantContainer{" +
                    "creatureEnchant=" + creatureEnchant +
                    ", level=" + level +
                    ", usedTo=" + usedTo +
                    '}';
        }
    }
}

File: src/main/java/hu/frontrider/arcana/capabilities/CreatureEnchantCapability.kt
package hu.frontrider.arcana.capabilities

import hu.frontrider.arcana.ThaumicArcana.MODID
import hu.frontrider.arcana.creatureenchant.backend.CreatureEnchant
import hu.frontrider.arcana.creatureenchant.backend.EnchantingBaseCircle
import net.minecraftforge.fml.common.registry.GameRegistry
import java.util.*

class CreatureEnchantCapability : ICreatureEnchant {
    override fun setEnabledStatus(enchant: CreatureEnchant, status: Boolean) {
        if(enabled.containsKey(enchant))
            enabled[enchant] = status
    }

    override var store: MutableMap<CreatureEnchant, CreatureEnchantContainer>
        get() = enchants
        set(value) {
            enchants = value
        }
    override var circle: EnchantingBaseCircle
        get() = enchantingBaseCircle
        set(value) {
            enchantingBaseCircle = value
        }

    override fun setEnabledStatus(store: MutableMap<CreatureEnchant, Boolean>) {
        enabled = store
    }

    override fun getEnabledStatus(): MutableMap<CreatureEnchant, Boolean> {
        return enabled
    }

    private var enchants: MutableMap<CreatureEnchant, CreatureEnchantContainer>
    private var enabled: MutableMap<CreatureEnchant, Boolean>

    private var enchantingBaseCircle: EnchantingBaseCircle

    init {
        enchants = HashMap()
        enabled = HashMap()
        enchantingBaseCircle = baseCircle!!
    }

    override fun hasEnchant(): Boolean {
        return enabled.values.stream().reduce(false) {
            buffer,current->
            buffer || current
        }
    }

    override fun hasEnchant(enchant: CreatureEnchant): Boolean {
        return when {
            !enabled.contains(enchant) -> false
            else -> enabled[enchant]!!
        }
    }

    override fun getLevel(enchant: CreatureEnchant): Int {
        return if (enchants[enchant] != null) enchants[enchant]!!.level else 0
    }

    override fun putEnchant(creatureEnchantContainer: CreatureEnchantContainer) {
        enchants[creatureEnchantContainer.creatureEnchant] = creatureEnchantContainer
        enabled[creatureEnchantContainer.creatureEnchant] = creatureEnchantContainer.enabled
    }


    data class CreatureEnchantContainer(val creatureEnchant: CreatureEnchant, val level: Int, val usedTo: Int, val enabled: Boolean = true)

    companion object {

        @GameRegistry.ObjectHolder("$MODID:normal")
        private var baseCircle: EnchantingBaseCircle? = null
    }
}
